### PR TITLE
Added transaction finalizer

### LINKED ISSUE #6955: resource service triggers background tasks before transaction is committed
The orchestration service calls the resource service's `mark_deployed` method [in a transaction context](https://github.com/inmanta/inmanta-core/blob/7cc942b19550f8ab20206752e6516bc0a23f01de/src/inmanta/server/services/orchestrationservice.py#L1151C49-L1151C49). The latter passes it on to `resource_action_update`, which adds a background task [for `ConfigurationModel.mark_done_if_done`](https://github.com/inmanta/inmanta-core/blob/7cc942b19550f8ab20206752e6516bc0a23f01de/src/inmanta/server/services/resourceservice.py#L1000) and [for `agent.resource_event`](https://github.com/inmanta/inmanta-core/blob/7cc942b19550f8ab20206752e6516bc0a23f01de/src/inmanta/server/services/resourceservice.py#L1011). These background tasks are added before the transaction is committed.

Consequences (hypothesis):
1. `mark_done_if_done` inspects resource state. I believe it now races (and typically loses) with the transaction commit so that it never sees all resources as deployed. The configuration model is therefore never marked as deployed. This impacts e.g. `inmanta-cli version list`. An easy reproduction is to simply run the quickstart. Export a new version without changes to the resources, then run `inmanta-cli  --host 172.30.0.3  version list -e SR_Linux`. It will remain in `deploying` until a repair is triggered. The web console does show all resources as deployed.
2. Does `agent.resource_event` inspect any database state (through an API call)? Is it safe to be used in this transaction context? We may be sending an event for something that hasn't committed yet. When the agent responds to the event by querying the server for resource state, it may get the wrong state.
3. There may or may not be other flows that pass a connection object into this method.

I did not go more in-depth at this time, because it is not entirely clear to me what is the best way forward. The `mark_done_if_done` approach might be no longer relevant, in which case it might be better to drop it and use something else for `inmanta-cli version list` rather than to address the concurrency problem. More input is needed from the team.

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Added transaction finalizer 
fixes #6955 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/server/services/orchestrationservice.py [hunk 1]
```diff
@@ -1067,93 +1068,97 @@ async def release_version(
...
-
-                # ensure we are the latest version
-                # this is required for the subsequent increment calculation to make sense
-                # this does introduce a race condition, with any OTHER release running concurrently on this environment
-                # We could lock the get_version_nr_latest_version for update to prevent this
-                if model.version < (latest_version or -1):
-                    raise Conflict(
-                        f"The version {version_id} on environment {env.id} "
-                        f"is older then the latest released version {latest_version}."
+                with ConnectionInTransaction(connection) as connection_holder:
```
Comment (wouterdb) [inline src/inmanta/server/services/orchestrationservice.py:1071]: indented one level

File: src/inmanta/server/services/resourceservice.py [hunk 2]
```diff
@@ -1005,23 +1006,33 @@ async def update_fields_resource(
...
-                aclient = self.agentmanager_service.get_agent_client(env.id, agent)
-                if aclient is not None:
-                    if change is None:
-                        change = const.Change.nochange
-                    self.add_background_task(
-                        aclient.resource_event(env.id, agent, resource_id, send_events, status, change, changes)
-                    )
+            def post_deploy_update() -> None:
+                assert model_version is not None  # mypy can't figure this out
+                # Make sure tasks are scheduled AFTER the tx is done
```
Comment (arnaudsjs) [inline src/inmanta/server/services/resourceservice.py:1016]:
```suggestion
                assert model_version is not None  # mypy can't figure this out
                # Make sure tasks are scheduled AFTER the tx is done.
                # This method is only called if the transaction commits successfully.
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/util/db.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        finalizer()
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        return None
+
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/util/db.py:62]:
```suggestion

```

File: src/inmanta/util/db.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    def call_after_tx(self, finalizer: Callable[[], object]) -> None:
+        finalizer()
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        return None
```
Comment (wouterdb) [inline src/inmanta/util/db.py:61]:
```suggestion
```

File: src/inmanta/util/db.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/util/db.py]: removed

**Review by arnaudsjs — APPROVED** [review]

**Pull request closed without merging**